### DISCUSSION TITLE
SIFTDescriptor non-square patch_size

### DISCUSSION THREAD

**Discussion by golnooshAbd** [kornia/kornia#1840] [Questions]
How is it possible to input non-square image to kornia.feature.SIFTDescriptor(patch_size=) ?
because  patch_size gets only int type, not tuple!
Thanks in advanced.

**Comment by ducha-aiki** [comment]
May I ask why do you need this? Usually SIFT is computed using some regular grid size and if you have rectangular grid, that means there are at least three options how to deal with that:

1. Using rectangular bins
2. Having bins of square size, but allowing different size bins (e.g. last would be smaller)
3. Resize patch to be squared.

As the SIFT paper does describe non-square bins, we have not implemented them either.


Right now it is possible to either resize patch to be square (option 3), or use DenseSIFT to implement (option 2) and then take the location, corresponding to the patch center:

```python3
import torch
import kornia.feature as KF
patch = torch.rand(1,1, 32,45)
hc = 32 //2
wc = 45 //2
denseSIFT = KF.DenseSIFTDescriptor()
out = denseSIFT(patch)
print(out.shape)
out_center = out[:,:,hc,wc]
print (out_center.shape)
```

Result:
```
torch.Size([1, 128, 32, 45])
torch.Size([1, 128])
```

Other options are not avaliable right now and you are welcomed to make a PR :)

**Reply by golnooshAbd** [reply]
I've got confused
I have 2 images with size of (1024, 512), and need a SIFT function for feature detection to find affine matrix parameter as a **classic** image registration between them, would you please help me how to use _kornia_ library?

I previously used opencv2 like this:
SIFT descriptor:
      
```
        sift = cv2.SIFT_create()
        kp, desc = sift.detectAndCompute(img, None)

        # Extract positions of key points
        kp = np.array([p.pt for p in kp]).T
```
     
`cv2.BFMatcher().knnMatch(desc_s, desc_t, k=2)` for matching between images
and RANSAC for parameters estimation.
Now I have to use torch-based functions to implement a classic image registration to find affine matrix.
Hope this help.

**Reply by ducha-aiki** [reply]
I see, you are talking about not SIFT descriptor, but SIFT as a whole detector and descriptor pipeline. 
In this case,  check this tutorial, which uses different local features - KeyNetAffNet, which are much better than SIFT.
https://kornia-tutorials.readthedocs.io/en/latest/image_matching_adalam.html

However, if you want use SIFT features instead, replace line `feature = KF.KeyNetAffNetHardNet(5000, True).eval().to(device)` to `feature = KF.SIFTFeature(5000, True).eval().to(device)`

This gives you keypoints and matches, similarly to opencv version
https://kornia.readthedocs.io/en/latest/feature.html#kornia.feature.SIFTFeature

Then you can use matches to run kornia.RANSAC https://kornia.readthedocs.io/en/v0.6.1/geometry.ransac.html 

We don't have an affine RANSAC implemented though, only homography and fundamental matrix though. You would need to linearize homography or implement affine RANSAC yourself (solver function, as the rest are the same).
